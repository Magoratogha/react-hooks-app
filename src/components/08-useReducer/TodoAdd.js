import { Fragment, memo } from "react";
import useForm from "../../hooks/useForm";
import PropTypes from 'prop-types';

const TodoAdd = memo(({ handleAddTodo }) => {
    const [ { description }, handleValueChange, resetFormValue ] = useForm({
        description: ''
    });

    const handleSubmit = (event) => {
        event.preventDefault();
        if (description.trim()) {
            const newTodo = {
                id: new Date().getTime(),
                description,
                isDone: false
            };
            handleAddTodo(newTodo);
            resetFormValue();
        }
    };

    return (
        <Fragment>
            <h4>Add todo</h4>
            <hr></hr>
            <form onSubmit={handleSubmit}>
                <input
                    className="form-control"
                    type="text"
                    name="description"
                    placeholder="To do..."
                    autoComplete="off"
                    value={description}
                    onChange={handleValueChange} />

                <div className="d-grid mt-2">
                    <button
                        disabled={!description.trim()}
                        type="submit"
                        className="btn btn-outline-dark btn-sm">
                        Add
                    </button>
                </div>
            </form>
        </Fragment>
    );
});

TodoAdd.propTypes = {
    handleAddTodo: PropTypes.func.isRequired
};

export default TodoAdd;