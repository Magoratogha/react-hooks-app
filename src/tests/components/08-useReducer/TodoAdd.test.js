import { render, screen } from "@testing-library/react";
import TodoAdd from "../../../components/08-useReducer/TodoAdd";
import userEvent from "@testing-library/user-event";

describe('Test on <TodoAdd />', () => {
    
    const user = userEvent.setup();
    const handleAddTodoMock = jest.fn();
    
    test('should render properly', () => {
        const { container } = render(<TodoAdd handleAddTodo={ handleAddTodoMock }/>);
        expect(container).toMatchSnapshot();
    });
    
    test('should not call handleAddTodo function if value is empty', async () => {
        render(<TodoAdd handleAddTodo={ handleAddTodoMock }/>);
        const submitButton = screen.queryByText('Add', { selector: 'button' });
        await user.click(submitButton);
        expect(handleAddTodoMock).not.toHaveBeenCalled();
    });
    
    test('should call handleAddTodo function if value is not empty', async () => {
        render(<TodoAdd handleAddTodo={ handleAddTodoMock }/>);
        const formInput = screen.queryByPlaceholderText('To do...', { exact: true });
        const submitButton = screen.queryByText('Add', { selector: 'button' });
        await user.type(formInput, 'Test');
        await user.click(submitButton);
        expect(handleAddTodoMock).toHaveBeenCalledWith({
            id: expect.any(Number),
            description: 'Test',
            isDone: false
        });
        expect(formInput).toHaveValue('');
    });
});
