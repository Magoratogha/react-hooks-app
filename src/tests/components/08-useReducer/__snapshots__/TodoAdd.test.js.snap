// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <TodoAdd /> should render properly 1`] = `
<div>
  <h4>
    Add todo
  </h4>
  <hr />
  <form>
    <input
      autocomplete="off"
      class="form-control"
      name="description"
      placeholder="To do..."
      type="text"
      value=""
    />
    <div
      class="d-grid mt-2"
    >
      <button
        class="btn btn-outline-dark btn-sm"
        disabled=""
        type="submit"
      >
        Add
      </button>
    </div>
  </form>
</div>
`;
